source/sigv4.c

```diff
@@ -1814,10 +1819,69 @@ static void generateCredentialScope( const SigV4Parameters_t * pSigV4Params,
 
 /*-----------------------------------------------------------*/
 
-    static SigV4Status_t setQueryStringFieldsAndValues( const char * pQuery,
-                                                        size_t queryLen,
-                                                        size_t * pNumberOfParameters,
-                                                        CanonicalContext_t * pCanonicalRequest )
+    static void processAmpersantInQueryString( bool fieldHasValue,
+                                               size_t currQueryIndex,
+                                               size_t startOfFieldOrValue,
+                                               size_t * pCurrParamCount,
+                                               const char * pQuery,
+                                               CanonicalContext_t * pCanonicalRequest )
+    {
+        bool previousParamIsValid = true;
+
+        assert( pCurrParamCount != NULL );
+        assert( pQuery != NULL );
+        assert( pCanonicalRequest != NULL );
+
+        /* Ignore unnecessary '&' that represent empty parameter names.
+         * Trimmable '&'s can be leading, trailing or repeated as separators between parameter
+         * pairs.
+         * For example, this function will prune "&p1=v1&&p2=v2&" to "p1=v1&p2=v2". */
+        if( !( fieldHasValue ) && ( ( ( currQueryIndex - startOfFieldOrValue ) == 0U ) ) )
+        {
+            /* Do nothing as previous parameter name is empty. */
+            previousParamIsValid = false;
+        }
+
+        /* Case when parameter has empty value even though the query parameter entry has the form
+         * "<param>=". */
+        else if( ( fieldHasValue ) && ( ( currQueryIndex - startOfFieldOrValue ) == 0U ) )
+        {
+            /* Store the previous parameter's empty value information. Use NULL to represent empty value. */
+            setQueryParameterValue( *pCurrParamCount, NULL, 0U, pCanonicalRequest );
+        }
+
+        /* Case when a new query parameter pair has begun without the previous parameter having an associated value,
+         * i.e. of the format "<param1>&<param2>=<value>" where "<param1>".
+         * In thus case, as the previous parameter does not have a value, we will store an empty value for it. */
+        else if( !( fieldHasValue ) )
```

Comment:
You can combine either the first `if` and `else if` into a single `if( ( currQueryIndex - startOfFieldOrValue ) == 0U )`, or the first `if` and this `else if ` into a single `if( !fieldHasValue )`

Reply:
I'm not sure that I follow your suggestion. All the four cases are distinct with separate handling. I have changed the if-else chain to instead use a nested if-else to be based off of the `fieldHasValue` data to make it easier to follow.

Reply:
Yep, that's what I meant